### PR TITLE
remove Application dependency from UrlDispatcher

### PR DESCRIPTION
**Pull request by fafhrd91**
I think UrlDispatcher should not depend on Application object. 
Here is backward compatible change to remove this dependency.

@asvetlov

### CODE REVIEW

**Comment by fafhrd91** [comment]
we need to update docs and add deprecation warning to `router.add_subapp`

i think `UrlDispatcher` should be part of public api

File: aiohttp/web.py [hunk 1]
```diff
@@ -125,6 +126,23 @@ def handler(app):
         reg_handler('on_shutdown')
         reg_handler('on_cleanup')
 
+    def add_subapp(self, prefix, subapp):
+        if self.frozen:
+            raise RuntimeError(
+                "Cannod add sub application to frozen application")
+        if subapp.frozen:
+            raise RuntimeError("Cannod add frozen application")
```
Comment (popravich) [inline aiohttp/web.py]: "cannod" typo

File: aiohttp/web_urldispatcher.py [hunk 2]
```diff
@@ -891,20 +883,25 @@ def add_delete(self, *args, **kwargs):
...
+        super().freeze()
+        for resource in self._resources:
+            resource.freeze()
+
+
+def _wrap_add_subbapp(app):
+
+    def add_subapp(prefix, subapp):
         if subapp.frozen:
             raise RuntimeError("Cannod add frozen application")
```
Comment (popravich) [inline aiohttp/web_urldispatcher.py]: ...a-ha, copied from here — "cannod"

Comment (fafhrd91) [reply aiohttp/web_urldispatcher.py]: good catch

**Comment by fafhrd91** [comment]
@asvetlov objections?

also what about `UrlDispatcher` as part of public api?

**Comment by asvetlov** [comment]
Hmm. I dislike `router.add_subapp = _wrap_add_subbapp(self)` -- looks like a hack.

Also I doubt if it will work with all possible router implementations.
Adding `add_subapp` to `Application` doesn't look reasonable -- we are adding a way to register sub applications only but not regular routes.

I suggest keeping the code as is.

**Comment by fafhrd91** [comment]
`router.add_subapp` just for backward compatibility until next release

**Pull request merged**